adapter/prometheus/prometheus.go

```diff
@@ -32,40 +38,185 @@ type (
+	for _, m := range metrics {
+		switch m.Kind {
+		case adapter.Gauge:
+			metricsMap[m.Name] = mustRegisterOrGet(newGaugeVec(m.Name, m.Description, m.Labels))
+		case adapter.Counter:
+			metricsMap[m.Name] = mustRegisterOrGet(newCounterVec(m.Name, m.Description, m.Labels))
+		default:
+			// TODO: should we return error here, instead of logging?
+			env.Logger().Warningf("unknown metric kind for: %v", m)
+		}
+	}
+
+	return &prom{metricsMap}, nil
 }
 
-func (*prom) Record([]adapter.Value) error {
-	return nil
+func (p *prom) Record(vals []adapter.Value) error {
+	var result *multierror.Error
+
+	for _, val := range vals {
+		collector, found := p.metrics[val.Name]
+		if !found {
+			result = multierror.Append(result, fmt.Errorf("could not find metric description for %s", val.Name))
+			continue
+		}
+		switch val.Kind {
+		case adapter.Gauge:
+			vec := collector.(*prometheus.GaugeVec)
+			amt, err := promValue(val)
+			if err != nil {
+				result = multierror.Append(result, fmt.Errorf("could get value for metric %s", val.Name))
+				continue
+			}
+			vec.With(promLabels(val.Labels)).Set(amt)
+		case adapter.Counter:
+			vec := collector.(*prometheus.CounterVec)
+			amt, err := promValue(val)
+			if err != nil {
+				result = multierror.Append(result, fmt.Errorf("could get value for metric %s", val.Name))
+				continue
+			}
+			vec.With(promLabels(val.Labels)).Add(amt)
+		default:
+			result = multierror.Append(result, fmt.Errorf("could not process metric value: %v", val))
+		}
+	}
+
+	return result.ErrorOrNil()
 }
 
 func (*prom) Close() error { return nil }
+
+func newCounterVec(name, desc string, labels map[string]adapter.LabelKind) *prometheus.CounterVec {
+	c := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: safeName(name),
+			Help: desc,
+		},
+		labelNames(labels),
+	)
+	return c
+}
+
+func newGaugeVec(name, desc string, labels map[string]adapter.LabelKind) *prometheus.GaugeVec {
+	c := prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: safeName(name),
+			Help: desc,
+		},
+		labelNames(labels),
+	)
+	return c
+}
+
+func labelNames(m map[string]adapter.LabelKind) []string {
+	i := 0
+	keys := make([]string, len(m))
+	for k := range m {
+		keys[i] = safeName(k)
+		i++
+	}
+	return keys
+}
+
+// borrowed from prometheus.MustRegisterOrGet. However, that method is
+// targeted for removal soon(tm). So, we duplicate that functionality here
+// to maintain the functionality, as we have a use case for the convenience.
+func mustRegisterOrGet(c prometheus.Collector) prometheus.Collector {
+	if err := prometheus.Register(c); err != nil {
+		if are, ok := err.(prometheus.AlreadyRegisteredError); ok {
+			return are.ExistingCollector
+		}
+		panic(err)
+	}
+	return c
+}
+
+func safeName(n string) string {
+	s := strings.TrimPrefix(n, "/")
+	return charReplacer.Replace(s)
+}
+
+func promValue(val adapter.Value) (float64, error) {
+	switch i := val.MetricValue.(type) {
```

Comment:
I feel like we can do a better job here given that we're the ones constructing the value (in the aspect manager). AFAIK the value must be produced from eval'ing some expression over the attributes, so I think the only types you need to worry about are those that can be produced from an evaluation: int64, float64, and maybe string.